### ISSUE TITLE
Phantom (deleted) host holds onto mon process, and cluster is degrated

### ISSUE THREAD

**Issue by skandragon** [rook/rook#13118]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

(see below for what I did to get here)

It appears I have a "host" in ceph even though it has been removed `kubectl delete node kube06`, and for some reason a mon is locked onto that node, and refused to stop going there.

These pods are stuck:

```
NAMESPACE                       NAME                                                     READY   STATUS      RESTARTS        AGE     IP              NODE       NOMINATED NODE   READINESS GATES
rook-ceph                       rook-ceph-mon-i-66f9cf78bb-vdfg4                         0/1     Pending     0               5m56s   <none>          <none>     <none>           <none>
rook-ceph                       rook-ceph-osd-prepare-kube02-f4r9h                       0/1     Pending     0               53m     <none>          <none>     <none>           <none>
rook-ceph                       rook-ceph-osd-prepare-kube04-8p9xv                       0/1     Pending     0               53m     <none>          <none>     <none>           <none>
```

**Expected behavior:**

* Deleted nodes should not get services locked to them.
* osd-prepare jobs should not get assigned to nodes that no longer exist.

**How to reproduce it (minimal and precise):**

I started a cluster several years ago with nodes called `kube01` through `kube06`, `kube10` through `kube13`, and `kube20` and `kube21`.  I wanted to remove the older, smaller nodes `kube01` through `kube06` so I followed these steps for each:

1.  Remove the node with `kubectl delete node`
2. Follow the instructions to remove the OSD by ID: `kubectl rook-ceph rook purge-osd` for the correct ID
3. Wait for things to settle
4. Move on to the next node

I deleted them in a slightly strange order:  kube02, kube01, kube03, kube04, kube05, and finally kube06.  Things worked without an issue and the OSDs are all deleted, and the storage has fully recovered.

However, those jobs (including the stuck mon) are node-locked, and I think that's because the host still appears in the ceph UI, and I assume this means it's in some other list that I cannot delete.

From what I have seen, I should use `ceph orch host rm` to remove a host manually, but:

```
% kubectl rook-ceph ceph orch host ls
Info: running 'ceph' command with args: [orch host ls]
HOST      ADDR                  LABELS  STATUS
kube10    10.45.101.1/kube10
kube11    10.45.101.2/kube11
kube12    10.45.101.3/kube12
kube13    10.45.101.4/kube13
kube20    10.45.101.0/kube20
kube21    10.45.15.255/kube21
kube-cp1  10.45.102.1/kube-cp1
kube-cp2  10.45.102.2/kube-cp2
kube-cp3  10.45.102.3/kube-cp3
9 hosts in cluster
```

```
% kubectl get nodes
NAME       STATUS   ROLES           AGE    VERSION
kube-cp1   Ready    control-plane   401d   v1.28.2
kube-cp2   Ready    control-plane   401d   v1.28.2
kube-cp3   Ready    control-plane   401d   v1.28.2
kube10     Ready    <none>          401d   v1.28.2
kube11     Ready    <none>          401d   v1.28.2
kube12     Ready    <none>          401d   v1.28.2
kube13     Ready    <none>          219d   v1.28.2
kube20     Ready    <none>          398d   v1.28.2
kube21     Ready    <none>          16h    v1.28.2
```

It's odd that the UI shows a host...  and I'm not sure how to get rid of it, but as it is shown with `mon: 1` and the kubernetes deployment tries to lock it to kube06...  I'm not sure what to do next.

(kube-cp* are control pane nodes and have no storage, and are not able to be scheduled)

(kube10, 11, 12, and 20 have storage, one OSD each currently)

![Screenshot 2023-10-28 at 4 03 50 AM](https://github.com/rook/rook/assets/43517/d4b74e69-b516-4fd0-a52c-90056ac119df)

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary

**Logs to submit**:

I've searched many logs but other than the operator attempting to wait for mon-i to come up, it is stuck.

* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

  To get logs, use `kubectl -n <namespace> logs <pod name>`
  When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
  Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Cluster Status to submit**:

* Output of kubectl commands, if necessary

  To get the health of the cluster, use `kubectl rook-ceph health`
  To get the status of the cluster, use `kubectl rook-ceph ceph status`
  For more details, see the [Rook kubectl Plugin](https://rook.io/docs/rook/latest-release/Troubleshooting/kubectl-plugin)

**Environment**:
* OS (e.g. from /etc/os-release): Talos metal
* Kernel (e.g. `uname -a`): talos
* Cloud provider or hardware configuration: talos
* Rook version (use `rook version` inside of a Rook Pod): 1.12.7
* Storage backend version (e.g. for ceph do `ceph -v`): `ceph version 17.2.3 (dff484dfc9e19a9819f375586300b3b79d80034d) quincy (stable)`
* Kubernetes version (use `kubectl version`): 1.28.2
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): talos metal
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)): `HEALTH_WARN 1/3 mons down, quorum c,j`

**Comment by skandragon** [comment]
I managed to get out of this state by:

1.  Manually edit the mon-i deployment to lock it to a different node.  This caused it to start, and the operator to be happy and continue doing its thing.  I don't know if this is a permanent fix in that if I remove that node, if it will try to move it back to -04 or not...
2. Manually remove the controlling jobs for the OSD jobs for hosts which no longer exist.

(1) managed to also remove kube-04 from the UI.

**Comment by sp98** [comment]
@skandragon 
- What message did you see in the pending pod description? (kubectl describe -n <namespace> <pending-pod-name>
- Was cleaning up the rook-ceph cluster the end goal here? If yes, we have [cleanup guide](https://rook.io/docs/rook/v1.12/Getting-Started/ceph-teardown/#cleaning-up-a-cluster).

**Comment by sp98** [comment]
> I managed to get out of this state by:
> 
> 1. Manually edit the mon-i deployment to lock it to a different node.  This caused it to start, and the operator to be happy and continue doing its thing.  I don't know if this is a permanent fix in that if I remove that node, if it will try to move it back to -04 or not...

This looks like an expected behavior to me. Most likely, you might have remove a node and no new node was added for the mon-i pod to back fail back to.  

Rook tries to failover mon-i to, say, mon-j, but since there is no node available, rook does not failover mon-i and remains in pending state.

**Comment by skandragon** [comment]
> @skandragon
> 
> * What message did you see in the pending pod description? (`kubectl describe -n namespace pending-pod-name`)

I'm not entirely sure, and I've since managed to clean up things so it reports healthy again.

> * Was cleaning up the rook-ceph cluster the end goal here? If yes, we have [cleanup guide](https://rook.io/docs/rook/v1.12/Getting-Started/ceph-teardown/#cleaning-up-a-cluster).

No, the purpose was to remove an OSD from a node, and then remove the node itself from Kubernetes entirely.   I was able to remove the OSD, but I had to do some additional cleanup to convince Ceph/Rook that the node didn't exist and should not get anything assigned to it.

**Comment by skandragon** [comment]
> This looks like an expected behavior to me. Most likely, a node was removed and no new node was added for the mon-i pod to fail back onto.

I did not add a new node, but there were 7 nodes remaining after I deleted the one, so there were hosts available.  The mon-i should have been assigned to one of them, but this did not happen automatically, and required manual intervention.

My guess here is something is asking for a host list, and that list includes where things are currently targeting, even if those services are not up.  Since that list includes defunct hosts from Rook/Ceph's POV, it doesn't see a need to change.  By forcing the issue, the list of hosts available no longer has that defunct node where mon-i used to live, and it is assigned to another already available node.
